sdk/pack/variables/variables.go

```diff
@@ -0,0 +1,156 @@
+package variables
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/nomad-pack/internal/pkg/errors/packdiags"
+	"github.com/hashicorp/nomad-pack/sdk/pack"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/mitchellh/go-wordwrap"
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/convert"
+)
+
+type ID string
+
+func (p ID) String() string { return string(p) }
+
+// Variable encapsulates a single variable as defined within a block according
+// to variableFileSchema and variableBlockSchema.
+type Variable struct {
+
+	// Name is the variable label. This is used to identify variables being
+	// overridden and during templating.
+	Name ID
+
+	// Description is an optional field which provides additional context to
+	// users identifying what the variable is used for.
+	Description    string
+	hasDescription bool
+
+	// Default is an optional field which provides a default value to be used
+	// in the absence of a user-provided value. It is only in this struct for
+	// documentation purposes
+	Default    cty.Value
+	hasDefault bool
+
+	// Type represents the concrete cty type of this variable. If the type is
+	// unable to be parsed into a cty type, it is invalid.
+	Type    cty.Type
+	hasType bool
+
+	// Value stores the variable value and is used when converting the cty type
+	// value into a Go type value.
+	Value cty.Value
+
+	// DeclRange is the position marker of the variable within the file it was
+	// read from. This is used for diagnostics.
+	DeclRange hcl.Range
+}
+
+func (v *Variable) SetDescription(d string) { v.Description = d; v.hasDescription = true }
+func (v *Variable) SetDefault(d cty.Value)  { v.Default = d; v.hasDefault = true }
+func (v *Variable) SetType(t cty.Type)      { v.Type = t; v.hasType = true }
+
+func (v *Variable) String() string { return asJSON(v) }
+
+func (v *Variable) Equal(ivp *Variable) bool {
+	if v == ivp {
+		return true
+	}
+	cv, ov := *v, *ivp
+	eq := cv.Name == ov.Name &&
+		cv.Description == ov.Description &&
+		cv.hasDescription == ov.hasDescription &&
+		cv.Default == ov.Default &&
+		cv.hasDefault == ov.hasDefault &&
+		cv.Type == ov.Type &&
+		cv.hasType == ov.hasType &&
+		cv.Value == ov.Value
+
+	return eq
+}
+
+func (v Variable) AsOverrideString(pID pack.ID) string {
```

Comment:
value receiver as opposed to all other methods being pointer receivers (vide James' comment above)